appbuilder/project/project-base.ts

```diff
@@ -1,29 +1,50 @@
+import {EOL} from "os";
 import Future = require("fibers/future");
-import { TARGET_FRAMEWORK_IDENTIFIERS } from "../../mobile/constants";
+import { TARGET_FRAMEWORK_IDENTIFIERS } from "../../constants";
 import * as path from "path";
-import { startPackageActivityNames } from "../../mobile/constants";
+import { startPackageActivityNames } from "../../constants";
+import * as helpers from "../../helpers";
+export abstract class ProjectBase implements Project.IProjectBase {
+	private static VALID_CONFIGURATION_CHARACTERS_REGEX = "[-_A-Za-z0-9]";
+	private static CONFIGURATION_FROM_FILE_NAME_REGEX = new RegExp(`^[.](${ProjectBase.VALID_CONFIGURATION_CHARACTERS_REGEX}+?)[.]abproject$`, "i");
 
-export class Project implements Project.IProjectBase {
-	constructor(private $cordovaProjectCapabilities: Project.ICapabilities,
-		private $fs: IFileSystem,
-		private $logger: ILogger,
-		private $nativeScriptProjectCapabilities: Project.ICapabilities,
-		private $projectConstants: Project.IConstants) { }
+	private _hasBuildConfigurations = false;
+	protected _shouldSaveProject = false;
+	protected _projectData: Project.IData;
+	public configurationSpecificData: IDictionary<Project.IData>;
+
+	constructor(protected $cordovaProjectCapabilities: Project.ICapabilities,
+		protected $errors: IErrors,
+		protected $fs: IFileSystem,
+		protected $logger: ILogger,
+		protected $nativeScriptProjectCapabilities: Project.ICapabilities,
+		protected $options: ICommonOptions,
+		protected $projectConstants: Project.IConstants,
+		protected $staticConfig: Config.IStaticConfig) {
 
-	public projectDir: string;
-	public getProjectDir(): IFuture<string> {
-		return Future.fromResult(this.projectDir);
+	}
+
+	// This property is purposely written as two separate methods so that only get/set can be overriden
+	protected getShouldSaveProject(): boolean {
+		return this._shouldSaveProject;
+	}
+
+	protected setShouldSaveProject(shouldSaveProject: boolean) {
+		this._shouldSaveProject = shouldSaveProject;
 	}
 
 	public get projectData(): Project.IData {
-		if(this.projectDir) {
-			let projectFile = path.join(this.projectDir, this.$projectConstants.PROJECT_FILE);
-			let jsonContent: Project.IData = this.$fs.readJson(projectFile).wait();
-			this.$logger.trace("Project data is: ", jsonContent);
-			return jsonContent;
-		}
+		this.readProjectData().wait();
+		return this._projectData;
+	}
 
-		return null;
+	public set projectData(projectData: Project.IData) {
+		this._projectData = projectData;
+	}
+
+	public projectDir: string;
+	public getProjectDir(): IFuture<string> {
```

Comment:
I think we can, but it'd have to be on a separate PR - this method is widely used throughout the code and a change like that would bring about a lot of unnecessary noise for this PR